File: algorithm/src/main/java/solutions/L25.java
package solutions;

public class L25 {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode();
        dummy.next = head;

        ListNode pre = dummy;

        while(pre != null){
            ListNode a = pre.next;
            ListNode b = pre.next;
            for(int i = 0; i < k - 1; i++){
                if(b == null)break;
                else{
                    b = b.next;
                }
            }
            if(b == null)break;
            if(a == b)break;

            ListNode tmp = b.next;
            ListNode x = a;

            ListNode y = a.next;
            a.next = null;
            while(true){
                ListNode z = y.next;
                y.next = x;
                x = y;
                y = z;
                if(x == b)break;
            }
            a.next = tmp;
            pre.next = b;
            pre = a;

        }
        return dummy.next;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(1);
        ListNode l2 = new ListNode(2);
        ListNode l3 = new ListNode(3);
        ListNode l4 = new ListNode(4);
        ListNode l5 = new ListNode(5);
        l1.next = l2;
        l2.next = l3;
        l3.next = l4;
        l4.next = l5;
        ListNode ans = new L25().reverseKGroup(l1, 4);
        while (ans != null){
            System.out.println(ans.val);
            ans = ans.next;
        }
    }
}

File: hotalg/wish/L25.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

"""
给你链表的头节点 head ，每 k 个节点一组进行翻转，请你返回修改后的链表。

k 是一个正整数，它的值小于或等于链表的长度。如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。

你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。
"""
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        len = 0
        p = head
        while p != None:
            len += 1
            p = p.next

        dummy = ListNode()
        p = dummy
        q = head
        dummy.next = head
        while len >= k:
            x,y,z = self.reverse(q, k)
            p.next = x
            p = y
            q = z
            len -= k
        if len > 0:
            p.next = q
        return dummy.next


    def reverse(self, x, k):
        if k == 1:
            return (x, x, x.next)
        dummy = ListNode()
        dummy.next = x
        if x == None or x.next == None:
            return
        else:
            y = x.next
            x.next = None
            u,v,w = self.reverse(y, k-1)
            y.next = x
            return (u,x,w)
